crates/astria-bridge-withdrawer/src/withdrawer/ethereum/watcher.rs

```diff
@@ -0,0 +1,493 @@
+    }
+}
+
+async fn watch_for_sequencer_withdrawal_events(
+    contract: AstriaWithdrawer<Provider<Ws>>,
+    event_tx: mpsc::Sender<(WithdrawalEvent, LogMeta)>,
+    from_block: u64,
+) -> Result<()> {
+    let events = contract
+        .sequencer_withdrawal_filter()
+        .from_block(from_block)
+        .address(contract.address().into());
+
+    let mut stream = events.stream().await.unwrap().with_meta();
+
+    while let Some(item) = stream.next().await {
+        if let Ok((event, meta)) = item {
+            event_tx
+                .send((WithdrawalEvent::Sequencer(event), meta))
+                .await
+                .wrap_err("failed to send sequencer withdrawal event; receiver dropped?")?;
+        } else if let Err(e) = item {
+            return Err(e).wrap_err("failed to read from event stream; event stream closed?");
+        }
+    }
+
+    Ok(())
+}
+
+async fn watch_for_ics20_withdrawal_events(
+    contract: AstriaWithdrawer<Provider<Ws>>,
+    event_tx: mpsc::Sender<(WithdrawalEvent, LogMeta)>,
+    from_block: u64,
+) -> Result<()> {
+    let events = contract
+        .ics_20_withdrawal_filter()
+        .from_block(from_block)
+        .address(contract.address().into());
+
+    let mut stream = events.stream().await.unwrap().with_meta();
+
+    while let Some(item) = stream.next().await {
+        if let Ok((event, meta)) = item {
+            event_tx
+                .send((WithdrawalEvent::Ics20(event), meta))
+                .await
+                .wrap_err("failed to send ics20 withdrawal event; receiver dropped?")?;
+        } else if let Err(e) = item {
+            return Err(e).wrap_err("failed to read from event stream; event stream closed?");
+        }
+    }
+
+    Ok(())
+}
+
+struct Batcher {
+    event_rx: mpsc::Receiver<(WithdrawalEvent, LogMeta)>,
+    batch_tx: mpsc::Sender<Batch>,
+    shutdown_token: CancellationToken,
+    fee_asset_id: asset::Id,
+    rollup_asset_denom: Denom,
+    asset_withdrawal_divisor: u128,
+}
+
+impl Batcher {
+    pub(crate) fn new(
+        event_rx: mpsc::Receiver<(WithdrawalEvent, LogMeta)>,
+        batch_tx: mpsc::Sender<Batch>,
+        shutdown_token: &CancellationToken,
+        fee_asset_id: asset::Id,
+        rollup_asset_denom: Denom,
+        asset_withdrawal_divisor: u128,
+    ) -> Self {
+        Self {
+            event_rx,
+            batch_tx,
+            shutdown_token: shutdown_token.clone(),
+            fee_asset_id,
+            rollup_asset_denom,
+            asset_withdrawal_divisor,
+        }
+    }
+}
+
+impl Batcher {
+    pub(crate) async fn run(mut self) -> Result<()> {
+        let mut curr_batch = Batch {
+            actions: Vec::new(),
+            rollup_height: 0,
+        };
+
+        loop {
+            select! {
+                () = self.shutdown_token.cancelled() => {
+                    info!("batcher shutting down");
+                    break;
+                }
+                item = self.event_rx.recv() => {
```

Comment:
This `select!` statement should break if `self.event_rx.recv() == None` because that means the channel is closed.